doc/man/mintupdate-cli.8
.\"	Title : mintupdate-cli
.\"	Author : gm10
.\"	February, 8 2019
.\"
.\" First parameter, NAME, should be all caps
.\" other parameters are allowed: see man(7), man(1)
.TH MINTUPDATE-CLI 8 "8 February 2019"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" for manpage-specific macros, see man(7)
.SH NAME
mintupdate-cli \- The Linux Mint Update Manager command line interface.

.SH SYNOPSIS
\fBmintupdate-cli\fR [options] {list | upgrade}

.SH DESCRIPTION
\fBmintupdate-cli\fR is the command line interface to mintupdate.
.PP
Unless the \fB-h\fR, or \fB--help\fR option is given, one of the commands below must be present:
.PP
\fBlist\fR
.RS 4
list available upgrades
.RE
.PP
\fBupgrade\fR
.RS 4
apply available upgrades
.RE

.SH OPTIONS
mintupdate-cli accepts the following options:
.TP
\fB-h\fR, \fB--help\fR
Show a help message and exit
.TP
\fB-k\fR, \fB--only-kernel\fR
Only include kernel updates
.TP
\fB-s\fR, \fB--only-security\fR
Only include security updates
.TP
\fB-i IGNORE\fR, \fB--ignore IGNORE\fR
List of updates to ignore (comma-separated list of source package names).

Note: You can also blacklist updates by adding them to /etc/mintupdate.blacklist, one source package per line.

To ignore a specific version, use the format package=version.
.TP
\fB-r\fR, \fB--refresh-cache\fR
Refresh the APT cache
.TP
\fB-d\fR, \fB--dry-run\fR
Simulation mode, don't upgrade anything
.TP
\fB-y\fR, \fB--yes\fR
Automatically answer yes to all questions and always install new configuration files (unless you also use "--keep-configuration" option).
.TP
\fB--install-recommends\fR
Install recommended packages regardless of system defaults.
.TP
\fB--keep-configuration\fR
Always keep local changes in configuration files (use with caution).
.TP
\fB-v\fR, \fB--version\fR
Display the current version

.SH "AUTHOR"
\fBThe Linux Mint developers\fR

.SH SEE ALSO
\fBmintupdate(8)\fR


debian/changelog
mintupdate (5.7.3) ulyssa; urgency=medium

  * rel_upgrade_root: Handle package additions/removals

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 09 Jan 2021 13:13:06 +0000

mintupdate (5.7.2) ulyssa; urgency=medium

  * l10n: Update translations
  * l10n: Update files

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 02 Jan 2021 16:48:27 +0000

mintupdate (5.7.1) ulyssa; urgency=medium

  * l10n: Update translations
  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 10 Dec 2020 12:00:12 +0000

mintupdate (5.7.0) ulyssa; urgency=medium

  [ Michael Webster ]
  * mintUpdate.py: Disable toolbar buttons when there are no updates available.

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 30 Nov 2020 16:41:24 +0000

mintupdate (5.6.9) ulyana; urgency=medium

  * l10n: Update translations
  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 24 Jun 2020 11:51:40 +0100

mintupdate (5.6.8) ulyana; urgency=medium

  * Simplify APT Cache Monitor

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 19 Jun 2020 17:09:23 +0100

mintupdate (5.6.7) ulyana; urgency=medium

  [ Butterfly ]
  * mintUpdate.py: Fixes #623 (#625)

  [ Clement Lefebvre ]
  * l10n: Set gettext and locale properly

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 17 Jun 2020 11:22:47 +0100

mintupdate (5.6.6) ulyana; urgency=medium

  [ felipecespedes ]
  * Show right description for grouped updates
  * avoid the try-except block for validating package in the cache

  [ NikoKrause ]
  * Change "Reboot required" icon in infobar

  [ Clement Lefebvre ]
  * tests: Remove call to obsolete get_eol_status()
  * CI: Target Mint 20
  * Revert "avoid the try-except block for validating package in the cache"
  * Revert "Show right description for grouped updates"

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 06 Jun 2020 15:14:42 +0100

mintupdate (5.6.5) ulyana; urgency=medium

  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 14 May 2020 08:22:58 +0100

mintupdate (5.6.4) ulyana; urgency=medium

  [ Michael Webster ]
  * systemd: Modify rules in service and timer files to be less impactful, but more reliable to users.

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 12 May 2020 14:52:12 +0100

mintupdate (5.6.3) ulyana; urgency=medium

  * l10n: Update translations
  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 12 May 2020 11:53:25 +0100

mintupdate (5.6.2) ulyana; urgency=medium

  [ Michael Webster ]
  * all: Get rid of some gtk deprecation warning spam.

  [ Nicolas Braud-Santoni ]
  * Order mintupdate-automation-upgrade after network-online

  [ Clement Lefebvre ]
  * Switch to symbolic tray icons
  * Update symbolic tray icons

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 07 May 2020 13:32:21 +0100

mintupdate (5.6.1) ulyana; urgency=medium

  [ Clement Lefebvre ]
  * CI: Use Mint 19.3

  [ JosephMcc ]
  * prefs: Add some margins around button in automation tab

  [ Butterfly ]
  * Fix: NameError: name 'ngettext' is not defined for Ulyana/Focal base

  [ Clement Lefebvre ]
  * Fix LSB call in Mint 20

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 21 Apr 2020 16:23:46 +0100

mintupdate (5.6.0) debbie; urgency=medium

  [ Michael Webster ]
  * mintUpdate.py: Disable the install button when no updates are selected (or available).
  * icons: Fix rendering of mintupdate-type-package-symbolic.svg.
  * icons: Lighten the outer ring of the symbolic 'package' icon to match previous appearance.

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 04 Feb 2020 15:25:05 +0000

mintupdate (5.5.9) tricia; urgency=medium

  * l10n: Update translations
  * l10n: Update files

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 06 Jan 2020 11:00:40 +0000

mintupdate (5.5.8) tricia; urgency=medium

  * mint-refresh-cache: Catch exceptions
  * l10n: Update translations
  * l10n: Generate files
  * Packaging: Depend on gir1.2-xapp-1.0

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 11 Dec 2019 14:28:19 +0000

mintupdate (5.5.7) tricia; urgency=medium

  * l10n: Update translations
  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 26 Nov 2019 17:35:01 +0000

mintupdate (5.5.6) tricia; urgency=medium

  * KernelWindow: Fix exception when kernel EOL tag isn't specified

 -- Clement Lefebvre <clem@mintbox3>  Wed, 13 Nov 2019 20:58:00 +0100

mintupdate (5.5.5) tricia; urgency=medium

  [ Clement Lefebvre ]
  * Indicator: Add support for tooltip

  [ Michael Webster ]
  * Simplify hiding/showing the main window, and make behavior more (#565)
  * mintUpdate.py: Fix visibility of menu items in the window.
  * mintUpdate.py: Add a way to terminate mintupdate from the window's file menu.

  [ Clement Lefebvre ]
  * Implement support for XAppStatusIcon (#567)
  * Remove timeshift check
  * Remove distribution EOL checks

  [ MrBesen ]
  * Fixed typo (#577)

  [ Michael Webster ]
  * mintUpdate.py: Improve reliability of window raising when clicking the status icon.

  [ Clement Lefebvre ]
  * l10n: Update POT

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 13 Nov 2019 13:04:11 +0100

mintupdate (5.5.4) tina; urgency=medium

  * Remove Python 3.6 code

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 05 Aug 2019 12:48:40 +0200

mintupdate (5.5.3) tina; urgency=medium

  * Kernel window: Fix exception when ACTIVE_KERNEL_VERSION isn't defined

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 03 Aug 2019 11:39:09 +0200

mintupdate (5.5.2) tina; urgency=medium

  * l10n: Update translations

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 28 Jul 2019 15:41:01 +0200

mintupdate (5.5.1) tina; urgency=medium

  * Change self-updating label
  * l10n: Update POT
  * Systray: Set tray icon before refresh thread

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 25 Jul 2019 10:34:55 +0200

mintupdate (5.5.0) tina; urgency=medium

  * Fix warning dialog not shown in kernel window
  * Kernelwindow: Remove unnecessary use of CSS style context

usr/lib/linuxmint/mintUpdate/mintupdate-cli.py
#!/usr/bin/python3

import argparse
import fnmatch
import os
import subprocess
import sys
import traceback

from checkAPT import APTCheck
from Classes import PRIORITY_UPDATES

if __name__ == "__main__":

    def is_blacklisted(blacklisted_packages, source_name, version):
        for blacklist in blacklisted_packages:
            if "=" in blacklist:
                (bl_pkg, bl_ver) = blacklist.split("=", 1)
            else:
                bl_pkg = blacklist
                bl_ver = None
            if fnmatch.fnmatch(source_name, bl_pkg) and (not bl_ver or bl_ver == version):
                return True
        return False

    parser = argparse.ArgumentParser(prog="mintupdate-cli")
    parser.add_argument("command", help="command to run (possible commands are: list, upgrade)")

    group = parser.add_mutually_exclusive_group()
    group.add_argument("-k", "--only-kernel", action="store_true", help="only include kernel updates")
    group.add_argument("-s", "--only-security", action="store_true", help="only include security updates")
    parser.add_argument("-i", "--ignore", help="list of updates to ignore (comma-separated list of source package names). Note: You can also blacklist updates by adding them to /etc/mintupdate.blacklist, one source package per line. To ignore a specific version, use the format package=version.")
    parser.add_argument("-r", "--refresh-cache", action="store_true", help="refresh the APT cache")
    parser.add_argument("-d", "--dry-run", action="store_true", help="simulation mode, don't upgrade anything")
    parser.add_argument("-y", "--yes", action="store_true", help="automatically answer yes to all questions and always install new configuration files (unless you also use \"--keep-configuration\" option)")
    parser.add_argument("--install-recommends", action="store_true", help="install recommended packages")
    parser.add_argument("--keep-configuration", action="store_true", default=False, help="always keep local changes in configuration files (use with caution)")
    parser.add_argument("-v", "--version", action="version", version="__DEB_VERSION__", help="Display the current version")

    args = parser.parse_args()
    try:
        if args.refresh_cache:
            subprocess.run("sudo /usr/bin/mint-refresh-cache", shell=True)
        check = APTCheck()
        check.find_changes()

        blacklisted = []
        if os.path.exists("/etc/mintupdate.blacklist"):
            with open("/etc/mintupdate.blacklist") as blacklist_file:
                for line in blacklist_file:
                    line = line.strip()
                    if line.startswith("#"):
                        continue
                    blacklisted.append(line)
        if args.ignore:
            blacklisted.extend(args.ignore.split(","))

        updates = []
        for source_name in sorted(check.updates.keys()):
            update = check.updates[source_name]
            if source_name in PRIORITY_UPDATES:
                updates.append(update)
            elif args.only_kernel and update.type != "kernel":
                continue
            elif args.only_security and update.type != "security":
                continue
            elif is_blacklisted(blacklisted, update.real_source_name, update.new_version):
                continue
            else:
                updates.append(update)

        if args.command == "list":
            for update in updates:
                print ("%-15s %-45s %s" % (update.type, update.source_name, update.new_version))
        elif args.command == "upgrade":
            packages = []
            for update in updates:
                packages += update.package_names
            arguments = ["apt-get", "install"]
            if args.dry_run:
                arguments.append("--simulate")
            if args.yes:
                environment = os.environ
                environment.update({"DEBIAN_FRONTEND": "noninteractive"})
                arguments.append("--assume-yes")
                if not args.keep_configuration:
                    arguments.extend(["--option", "Dpkg::Options::=--force-confnew"])
            else:
                environment = None
            if args.install_recommends:
                arguments.append("--install-recommends")
            if args.keep_configuration:
                arguments.extend(["--option", "Dpkg::Options::=--force-confold"])
            subprocess.call(arguments + packages, env=environment)
    except:
        traceback.print_exc()
        sys.exit(1)
